accepted/00080-api-versioning.md

```diff
@@ -0,0 +1,295 @@
+
+   # Non-existing handler
+   <Fault -1: 'The specified handler cannot be found'>
+   ```
+   Note: This fault code is
+   not reserved solely to report the API calling mismatch cases, but
+   is used in other cases (see the`CustomInfoHandler.java` file in
+   Uyuni).
+
+## Existing approach and problems
+
+We must provide a sane way for users to write robust scripts that
+target various Uyuni/SUMA versions.
+
+In the `spacecmd` tool, this is done by API version check endpoint
+(`api.getVersion` above). The logic simply checks the API version and
+reacts accordingly (e.g. warns the user, that given operation is not
+supported by the API version).
+
+A similar approach is taken by the `errata-import.pl` script by Steve
+Meier
+[1](https://github.com/stevemeier/cefs/blob/master/errata-import.pl). The
+script contains a list of supported versions and checks on runtime,
+whether the API version of the Uyuni/SUMA server is contained in this
+list.
+
+This works well, until the user needs to target both Uyuni and SUSE
+Manager in their scripts - the API version numbers are not consistent
+(API version `25` in Uyuni is currently not the same as version `25`
+in SUMA).
+
+TODO: Remove this probably. This described a buggy scenario.
+Second problem of this approach emerges, when the product developers
+forget to bump the API version on breaking changes. In this case, even
+a client application that solely relies on checking the API version
+can use an API method in a wrong way.
+
+Tackling these problems has various solutions, described in the
+[Solution proposals](#solution-proposals) section below.
+
+
+## Note on breaking and non-breaking changes
+
+The API mutates over time, in general there are 3 types of changes:
+
+1. growing (not breaking): adding a new namespace, introducing a new
+method in an existing namespace
+2. shrinking (breaking): removing a namespace, removing a method from
+an existing namespace
+3. modifying (potentially breaking):
+    1. breaking: changing a signature of an existing method
+    2. breaking: changing a behavior of an existing method
+    3. non breaking: adding a new field in a structure accepted /
+       returned by a method
+    4. breaking: removing a field in a structure accepted by a method /
+       returned by a method
+
+
+## Solution proposals
+
+The following section contains a list of solutions to the problems
+mentioned above. Some of the the suggestions are only hypothetical,
+but are included anyway, for the sake of completeness.
+
+
+### Solution 1
+
+- Treat the API versions independent in Uyuni and SUSE Manager. These
+  are 2 different projects and version `X` has a different contents
+  and compatibility in Uyuni and in SUMA.
+- Introduce an API "flavor" under a new `api.getFlavor` method, which
+  returns the project name (e.g. `uyuni`/`suse-manager`) read from a
+  config file.
+- Only "bump" the API versions on breaking changes.
```

Comment:
I think this RFC should propose a solution to the general problem of client scripts which want to differentiate their behavior based on any change, breaking or non-breaking (especially: the existence of a certain method that was added at some point in time).

I think we can already discard solutions that do not address this somehow.

Reply:
I suggested the definitive solution in the RFC comment, but that shouldn't be the way of doing this. Let's do it this way: I'll update the RFC and present the definitive solution as solution # 1 and we can iterate on that.

Reply:
Right - RFCs should propose one solution - the one the author thinks is best and wants to propose to the community.

Known alternatives should be listed (and, depending on context, explained at the appropriate level of detail) to inform readers that other possibilities have been explored and their pros/cons evaluated.